### PR TITLE
fix: Handle bad permission errors for bigquery test connections

### PR DESCRIPTION
**Pull request by hughhhh**
### SUMMARY

Whenever a user is testing a connection via bigquery and the current user doesn't have the proper permissions. Superset throws generic error vs. specifically informing the user that their permission needs to be updated. This PR leverages sip-40 and sip-41 work to properly inform the user specific error

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/superset/databases/commands/test_connection.py", line 64, in run
    if not engine.dialect.do_ping(conn):
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 560, in do_ping
    cursor.execute(self._dialect_specific_select_one)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/dbapi/_helpers.py", line 258, in with_closed_check
    return method(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 185, in execute
    formatted_operation, job_config=config, job_id=job_id
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 2776, in query
    query_job._begin(retry=retry, timeout=timeout)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/job.py", line 3179, in _begin
    super(QueryJob, self)._begin(client=client, retry=retry, timeout=timeout)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/job.py", line 646, in _begin
    timeout=timeout,
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 641, in _call_api
    return call()
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func
    on_error=on_error,
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 184, in retry_target
    return target()
  File "/usr/local/lib/python3.7/site-packages/google/cloud/_http.py", line 435, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.Forbidden: 403 POST <curl URL>
: User does not have bigquery.jobs.create permission in project <project_id>
```

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/databases/commands/exceptions.py [hunk 1]
```diff
@@ -131,7 +134,7 @@ class DatabaseTestConnectionDriverError(CommandInvalidError):
 
 
 class DatabaseTestConnectionUnexpectedError(CommandInvalidError):
-    message = _("Unexpected error occurred, please check your logs for details")
+    message = _(TEXT_DATABASE_CONN["database_test_connection_unexpected"])
```
Comment (betodealmeida) [inline superset/databases/commands/exceptions.py:137]: I'm not super familiar with gettext, but I think you need to move the `__()` call to `superset/config.py` where the string is defined, you can't use it with variables.

Comment (hughhhh) [reply superset/databases/commands/exceptions.py:137]: do you know what the `_()` assuming it is for translations right?

Comment (eschutho) [reply superset/databases/commands/exceptions.py:137]: from flask_babel import lazy_gettext as _

File: superset/errors.py [hunk 2]
```diff
@@ -173,6 +176,12 @@ class SupersetErrorType(str, Enum):
             ),
         },
     ],
+    SupersetErrorType.TEST_CONNECTION_DATABASE_PERMISSIONS_ERROR: [
+        {
+            "code": 1014,
```
Comment (AAfghahi) [inline superset/errors.py:181]: I think this is code 1017? 
```suggestion
            "code": 1017,
```
https://github.com/apache/superset/pull/14172/files

File: superset/errors.py [hunk 3]
```diff
@@ -173,6 +176,12 @@ class SupersetErrorType(str, Enum):
...
+            "message": _("Issue 1014 - User doesn't have the proper permissions"),
```
Comment (AAfghahi) [inline superset/errors.py:182]:
```suggestion
            "message": _("Issue 1017 - User doesn't have the proper permissions"),
```

File: superset/db_engine_specs/bigquery.py [hunk 4]
```diff
@@ -20,16 +20,23 @@
...
 
 from superset.db_engine_specs.base import BaseEngineSpec
+from superset.errors import SupersetErrorType
 from superset.utils import core as utils
 
 if TYPE_CHECKING:
     from superset.models.core import Database  # pragma: no cover
 
 
+INVALID_USER_PERMISSIONS = re.compile(
```
Comment (AAfghahi) [inline superset/db_engine_specs/bigquery.py:35]:
```suggestion
INVALID_USER_PERMISSIONS _REGEX= re.compile(
```

Comment (AAfghahi) [reply superset/db_engine_specs/bigquery.py:35]: though thinking about it, this should maybe be:
CONNECTION_DATABASE_PERMISSIONS_REGEX in order to mirror the error?

File: superset/db_engine_specs/bigquery.py [hunk 5]
```diff
@@ -86,6 +93,13 @@ class BigQueryEngineSpec(BaseEngineSpec):
         "P1Y": "{func}({col}, YEAR)",
     }
 
+    custom_errors = {
+        INVALID_USER_PERMISSIONS: (
+            __("bad permissions"),
```
Comment (AAfghahi) [inline superset/db_engine_specs/bigquery.py:98]:
```suggestion
            __("User does not have necessary permissions"),
```
How is this wording?

File: docs/src/pages/docs/Miscellaneous/issue_codes.mdx [hunk 6]
```diff
@@ -167,11 +168,18 @@ Either the database is spelled incorrectly or does not exist.
...
 
 The schema was either removed or renamed. Check that the schema is typed correctly and exists.
+
+## Issue 1017
+
+```
+We were unable to connect to your database. Please confirm that your service account has the Viewer and Job User roles on the project.
+```
+
+The user doesn't have the proper permissions to connect to the database
```
Comment (betodealmeida) [inline docs/src/pages/docs/Miscellaneous/issue_codes.mdx:185]: The convention we've been using is:

```
    ```
    Short message describing the problem.
    ```

    Descriptive message with actionable items.
```

So in this case it would be better to reverse the 2 messages you have here.

File: superset/errors.py [hunk 7]
```diff
@@ -47,6 +47,7 @@ class SupersetErrorType(str, Enum):
     CONNECTION_HOST_DOWN_ERROR = "CONNECTION_HOST_DOWN_ERROR"
     CONNECTION_ACCESS_DENIED_ERROR = "CONNECTION_ACCESS_DENIED_ERROR"
     CONNECTION_UNKNOWN_DATABASE_ERROR = "CONNECTION_UNKNOWN_DATABASE_ERROR"
+    CONNECTION_DATABASE_PERMISSIONS_ERROR = "TEST_CONNECTION_DATABASE_PERMISSIONS_ERROR"
```
Comment (betodealmeida) [inline superset/errors.py:50]: You also need to add this to:

- `superset-frontend/src/components/ErrorMessage/types.ts`
- `superset-frontend/src/setup/setupErrorMessages.ts`
- `docs/src/pages/docs/Miscellaneous/issue_codes.mdx`

File: superset/db_engine_specs/bigquery.py [hunk 8]
```diff
@@ -20,16 +20,23 @@
...
 from superset.db_engine_specs.base import BaseEngineSpec
+from superset.errors import SupersetErrorType
 from superset.utils import core as utils
 
 if TYPE_CHECKING:
     from superset.models.core import Database  # pragma: no cover
 
 
+CONNECTION_DATABASE_PERMISSIONS_REGEX = re.compile(
+    "(?:^|(?<= ))(403|POST|permission)(?:(?= )|$)"
```
Comment (betodealmeida) [inline superset/db_engine_specs/bigquery.py:36]: I think this is too broad, it might catch other errors that are unrelated.

It would be better to catch something like:

```python
"Access Denied: Project User does not have bigquery.jobs.create permission in project (?P<project>.+?)"
```

File: superset/db_engine_specs/bigquery.py [hunk 9]
```diff
@@ -86,6 +93,17 @@ class BigQueryEngineSpec(BaseEngineSpec):
         "P1Y": "{func}({col}, YEAR)",
     }
 
+    custom_errors = {
+        CONNECTION_DATABASE_PERMISSIONS_REGEX: (
+            __(
+                "We were unable to connect to your database. Please "
+                "confirm that your service account has the Viewer "
+                "and Job User roles on the project."
```
Comment (betodealmeida) [inline superset/db_engine_specs/bigquery.py:101]: This should also be simple, just describing the error. The actionable item should be in the issue code.

File: superset/errors.py [hunk 10]
```diff
@@ -213,6 +214,12 @@ class SupersetErrorType(str, Enum):
             ),
         }
     ],
+    SupersetErrorType.CONNECTION_DATABASE_PERMISSIONS_ERROR: [
+        {
+            "code": 1017,
+            "message": _("Issue 1017 - User doesn't have the proper permissions"),
```
Comment (betodealmeida) [inline superset/errors.py:220]:
```suggestion
            "message": _("Issue 1017 - User doesn't have the proper permissions."),
```

**Review by betodealmeida — CHANGES_REQUESTED** [review]
Hugh, this looks great, I just left a few comments to have it consistent with the format we've been using with other errors messages.

File: superset/db_engine_specs/bigquery.py (hunk 9 above)
Comment (hughhhh) [reply superset/db_engine_specs/bigquery.py:101]: This is the message sophie wanted specifically for user permissions

**Review by betodealmeida — APPROVED** [review]

File: superset/errors.py (hunk 7 above)
Comment (etr2460) [inline superset/errors.py:50]: typically it's good to keep the enum name and the value consistent. It's easier to track throughout the code
